src/query/src/dist_plan/planner.rs

```diff
@@ -41,16 +66,115 @@ impl ExtensionPlanner for DistExtensionPlanner {
         let maybe_merge_scan = { node.as_any().downcast_ref::<MergeScanLogicalPlan>() };
         if let Some(merge_scan) = maybe_merge_scan {
             if merge_scan.is_placeholder() {
-                let input = merge_scan.input().clone();
+                // ignore placeholder
                 planner
-                    .create_physical_plan(&input, session_state)
+                    .create_physical_plan(merge_scan.input(), session_state)
                     .await
                     .map(Some)
             } else {
-                Err(DataFusionError::NotImplemented("MergeScan".to_string()))
+                // TODO(ruihang): generate different execution plans for different variant merge operation
+                let input_plan = merge_scan.input();
+                let Some(table_name) = self.get_table_name(input_plan)? else {
+                    // no relation found in input plan, going to execute them locally 
+                    return planner
+                        .create_physical_plan(input_plan, session_state)
+                        .await
+                        .map(Some);
+                };
+                let input_schema = input_plan.schema().clone();
+                let substrait_plan: Bytes = DFLogicalSubstraitConvertor
+                    .encode(input_plan.clone())
+                    .context(error::EncodeSubstraitLogicalPlanSnafu)?
+                    .into();
+                let peers = self.get_peers(&table_name).await;
+                match peers {
+                    Ok(peers) => {
+                        let exec = MergeScanExec::new(
+                            table_name,
+                            peers,
+                            substrait_plan,
+                            Arc::new(input_schema.as_ref().into()),
+                            self.clients.clone(),
+                        );
+
+                        Ok(Some(Arc::new(exec) as _))
+                    }
+                    Err(_) => planner
+                        .create_physical_plan(input_plan, session_state)
+                        .await
+                        .map(Some),
+                }
             }
         } else {
             Ok(None)
         }
     }
 }
+
+impl DistExtensionPlanner {
+    /// Extract table name from logical plan
+    fn get_table_name(&self, plan: &LogicalPlan) -> Result<Option<TableName>> {
+        let mut extractor = TableNameExtractor::default();
+        plan.visit(&mut extractor)?;
+        Ok(extractor.table_name)
+    }
+
+    async fn get_peers(&self, table_name: &TableName) -> Result<Vec<Peer>> {
+        self.partition_manager
+            .find_table_region_leaders(table_name)
```

Comment:
It occurs to me that the MergeScan is doing an all regions scan? Does it use partition rule to find the target regions?

Reply:
No. It broadcast the sub-plan to all regions. This can be an optimization but I haven't took it into consideration for this time.

Reply:
todo here and we are good to go